io/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -728,27 +730,30 @@ object IO {
    * Submerges the error case of a disjunction into the `IO`. The inverse
    * operation of `IO.attempt`.
    */
-  final def absolve[E, A](v: IO[E, E \/ A]): IO[E, A] =
-    v.flatMap {
-      case -\/(e) => IO.fail(e)
-      case \/-(a) => IO.now(a)
-    }
+  final def absolve[E, A, B](f: FromEA[E, A] => B => IO[E, A])(v: IO[E, B]): IO[E, A] =
+    v.flatMap(f(FromEA[E, A](IO.fail, IO.now)))
+
+  @inline
+  private def absolveF[E, A](sinner: IO[E, FromEA[E, A] => IO[E, A]]) =
+    IO.absolve[E, A, FromEA[E, A] => IO[E, A]](fea => f => f(fea))(sinner)
+
+  @inline
+  private def absolveEither[E, A](sinner: IO[E, Either[E, A]]) =
+    IO.absolve[E, A, Either[E, A]](fea => b => b.fold(fea.err, fea.succ))(sinner)
 
   /**
    * Retrieves the supervisor associated with the fiber running the action
    * returned by this method.
    */
-  def supervisor[E]: IO[E, Throwable => IO[Void, Unit]] = new Supervisor()
+  def supervisor[E]: IO[E, Throwable => IO[Nothing, Unit]] = new Supervisor()
 
   /**
    * Requires that the given `IO[E, Maybe[A]]` contain a value. If there is no
    * value, then the specified error will be raised.
    */
-  final def require[E, A](error: E): IO[E, Maybe[A]] => IO[E, A] =
-    (io: IO[E, Maybe[A]]) => io.flatMap(Maybe.maybe(IO.fail[E, A](error))(IO.now[E, A]))
+  final def require[E, A](error: E): IO[E, MaybeIO[E, A]] => IO[E, A] =
```

Comment:
I'd do the same trick you used to eliminate `Maybe`, not this version (since `MaybeIO` should be deleted).

Reply:
The other case was a `Maybe` on the other end, it was returned not consumed, I can't use same trick here.

Reply:
Ah, then use `Disjunction[Unit, B]`, which is isomorphic to a `Maybe[B]`.